Add data retention policy (#188)
dandi/dandi-hub: doc/design/data-retention-policy.md

# Data Retention Policy

Dandihub data storage on AWS EFS is expensive, and we suppose that significant portions of the data
currently stored are no longer used. Data migration is where the cost becomes extreme.

> 

Reviewer: ```suggestion
Based on the policy outlined below, we will reset home directories after 45 days of inactivity to reduce Dandihub storage costs.
Additionally, we will calculate home directory usage to assess the effectiveness of this policy.

```

# Data Retention Policy

Dandihub data storage on AWS EFS is expensive, and we suppose that significant portions of the data
currently stored are no longer used. Data migration is where the cost becomes extreme.

## Persistent Data locations

Each user has access to 2 locations: `/home/{user}` and `/shared/`

> (Within jupyterhub `/home/{user}` the user always sees `/home/jovyan`, but is stored in EFS as their GitHub

Author: ```suggestion
Within jupyterhub `/home/{user}` the user always sees `/home/jovyan`, but is stored in EFS as their GitHub
```

> username.

Reviewer 2: ```suggestion
username).
```
or why bother with `()` at all here?

Author: Removed paren

## Known cache file cleanup 

We should be able to safely remove the following:
 - `/home/{user}/.cache`
 - `nwb_cache` 
 - Yarn Cache
 - `__pycache__`

>  - pip cache

Reviewer 2: In case user is still active -- I think it would be useful to report to the long running users, after reaching some threshold on any of those folders (e.g. 50MB) asking to clean them up.

Reviewer: Hi @asmacdo, should we add a separate point here about monitoring and reporting the quotas of cache directories?

## Determining Last Access

> EFS does not store metadata for the last access of the data. (Though they must track somehow to move
> to `Infrequent Access`)

Author: ```suggestion
EFS does not store metadata for the last access of the data. (Though they must track somehow to move to `Infrequent Access`)
```

Alternatives: 
 - use the [jupyterhub REST API](https://jupyterhub.readthedocs.io/en/stable/reference/rest-api.html#operation/get-users) check when user last used/logged in to hub.
 - dandiarchive login information

## Automated Data Audit

At some interval (30 days with no login?):

>    - find files larger than 1GB and mtime > 30 (?) days -- get total size and count

Author: ```suggestion
   - find files larger than 1 (?) GB and mtime > 30 (?) days -- get total size and count
```

>    - find files larger than 1 (?) GB and mtime > 30 (?) days -- get total size and count

Reviewer 2: might be worth even adding more aggresive
```suggestion
   - find files larger than 100 (?) GB and mtime > 10 (?) days -- get total size and count
   - find files larger than 1 (?) GB and mtime > 30 (?) days -- get total size and count
```

   - find _pycache_ and nwb-cache folders and pip cache and mtime > 30? days -- total sizes and list of them

Notify user if:

>    - total du exceeds some threshold (e.g. 100G)

Reviewer 2: ```suggestion
   - any of the above listed thresholds were reached
   - total du exceeds some threshold (e.g. 100G)
```

   - total outdated caches size exceeds some threshold (e.g. 1G)
   - prior notification was sent more than a week ago

Notification information:
   - large file list
   - summarized data retention policy
   - Notice number

>    - request to cleanup

Reviewer 2: meanwhile it might be worth creating a simple data record schema to store those records as well so they could be reused by the tools to assemble higher level stats etc.

### Non-response cleanup 

If a user has not logged in for 60 days (30 days initial + 30 days following audit), send a warning: 
`In 10 days the following files will be cleaned up`

> If the user has not logged in for 60 days (30 initial + 30 after audit + 10 warning):

Author: ```suggestion
If the user has not logged in for 70 days (30 initial + 30 after audit + 10 warning):
```

# Data Retention Policy

Dandihub data storage on AWS EFS is expensive, and we suppose that significant portions of the data

> currently stored are no longer used. Data migration is where the cost becomes extreme.

Reviewer: Since S3 buckets can now mount to EC2 instances (reference: [August 2023 blog post](https://aws.amazon.com/blogs/aws/mountpoint-for-amazon-s3-generally-available-and-ready-for-production-workloads/)) and S3 costs are ~10X cheaper than EFS, as part of this data retention work perhaps we should also look into what it would take to move to S3 storage (and discuss any features that would not be available with this migration)?

Reviewer: Filed #198

> ## Persistent Data locations

Reviewer: ```suggestion
## Persistent Data Locations
```

Each user has access to 2 locations: `/home/{user}` and `/shared/`.

Within jupyterhub `/home/{user}` the user always sees `/home/jovyan`, but is stored in EFS as their GitHub

> username.

Reviewer: We were previously considering providing a `/scratch` directory for each user that is automatically cleaned up after 30 days.  In addition to the policy for the `/home/jovyan` directory, do we also want to implement a `/scratch` directory with a 30 day clean up policy?

> ## Known cache file cleanup 

Reviewer: ```suggestion
## Known Cache File Cleanup
```

> We should be able to safely remove the following:

Reviewer: ```suggestion
When migrating data for new DandiHub deployments, the following cache directories can be safely removed:
```

Reviewer: On second thought, let's ignore this suggestion.

> ## Known cache file cleanup 
> 
> We should be able to safely remove the following:
>  - `/home/{user}/.cache`
>  - `nwb_cache` 
>  - Yarn Cache
>  - `__pycache__`
>  - pip cache

Reviewer: ```suggestion
```

## Determining Last Access

> EFS does not store metadata for the last access of the data. (Though they must track somehow to move to `Infrequent Access`)
> 
> Alternatives: 
>  - use the [jupyterhub REST API](https://jupyterhub.readthedocs.io/en/stable/reference/rest-api.html#operation/get-users) check when user last used/logged in to hub.
>  - dandiarchive login information

Reviewer: ```suggestion
Use the [JupyterHub REST API](https://jupyterhub.readthedocs.io/en/stable/reference/rest-api.html#operation/get-users) check when user last logged in to the hub..
```

## Automated Data Audit

> 

Reviewer: ```suggestion

At an interval of 7 days:
- Determine if total du exceeds 1 TB

```

Reviewer: I suggested a quota of 1 TB for home directories as many datasets are getting to be quite large.  This would provide temporary, high-capacity storage, but hopefully users won't get anywhere near this threshold.  This would cost $300/user/month for standard EFS, and $23/user/month if we move to Standard S3.

If we implement a scratch directory, then perhaps the home directory can have a much smaller quota.

Reviewer: This can be for our internal book keeping.

> At some interval (30 days with no login?):

Reviewer: ```suggestion
At an interval of 30 days with no login to JupyterHub:
```

>    - find files larger than 100 (?) GB and mtime > 10 (?) days -- get total size and count
>    - find files larger than 1 (?) GB and mtime > 30 (?) days -- get total size and count
>    - find _pycache_ and nwb-cache folders and pip cache and mtime > 30? days -- total sizes and list of them

Reviewer: ```suggestion
   - find files larger than 100 GB and mtime > 10 days -- get total size and count
   - find files larger than 1 GB and mtime > 30 days -- get total size and count
   - find _pycache_ and nwb-cache folders and pip cache and mtime > 30 days -- total sizes and list of them
```

> At some interval (30 days with no login?):
>    - find files larger than 100 (?) GB and mtime > 10 (?) days -- get total size and count
>    - find files larger than 1 (?) GB and mtime > 30 (?) days -- get total size and count
>    - find _pycache_ and nwb-cache folders and pip cache and mtime > 30? days -- total sizes and list of them

Reviewer: ```suggestion
```

Notify user if:
   - any of the above listed thresholds were reached

>    - total du exceeds some threshold (e.g. 100G)

Reviewer: I moved this threshold to the list above, increased it to 1 TB, and suggest that we monitor it on a more frequent basis of every 7 days (regardless of last login).

```suggestion
```

>    - total du exceeds some threshold (e.g. 100G)

Reviewer: ```suggestion
   - total home directory disk usage exceeds 1 TB
```

Reviewer: I suggested a quota of 1 TB for home directories as many datasets are getting to be quite large. This would provide temporary, high-capacity storage, but hopefully users won't get anywhere near this threshold. This would cost $300/user/month for standard EFS, and $23/user/month if we move to Standard S3.

If we implement a scratch directory, then perhaps the home directory can have a much smaller quota.

>    - total outdated caches size exceeds some threshold (e.g. 1G)

Reviewer: ```suggestion
   - total outdated caches size exceeds 1 GB
```

   - prior notification was sent more than a week ago

Notification information:

>    - large file list

Reviewer: ```suggestion
    - summarized audit data (total size and count for each of the above thresholds)
    - large file list
```

> Notify user if:
>    - any of the above listed thresholds were reached
>    - total du exceeds some threshold (e.g. 100G)
>    - total outdated caches size exceeds some threshold (e.g. 1G)
>    - prior notification was sent more than a week ago
> 
> Notification information:
>    - large file list
>    - summarized data retention policy
>    - Notice number
>    - request to cleanup

Reviewer: Based on our discussion this week, for now let's just reset home directories if the users haven't accessed the hub in 45 days.  If we have many users storing more than 10 GB over the long-run we can add pieces of this back.

```suggestion
```

> ### Non-response cleanup 
> 
> If a user has not logged in for 60 days (30 days initial + 30 days following audit), send a warning: 
> `In 10 days the following files will be cleaned up`
> 
> If the user has not logged in for 70 days (30 initial + 30 after audit + 10 warning):
> `The following files were removed`
> 
> Reset timer. 

Reviewer: ```suggestion
### Non-response cleanup 

If a user has not logged in for 30 days, send a warning: 
`In 15 days, the files in your home directory on DANDI Hub will be deleted.  Please review your files stored on DANDI Hub and upload any relevant files to your respective Dandisets on DANDI Archive.`

If the user has not logged in for 45 days, send a confirmation:
`The files in your home directory on DANDI Hub were deleted.`
```
